File: lib/download/download_manager.dart
import 'dart:async';

import '../src/rust/api/download_manager.dart' as frb;
import '../src/rust/download/download_status.dart' as frb_status;

export '../src/rust/download/download_status.dart';

/// Download task snapshot from the Rust download pool.
class DownloadTask {
  DownloadTask({
    required this.id,
    required this.uri,
    required this.priority,
    required this.progress,
    required this.cachedBytes,
    required this.downloadedBytes,
    required this.totalBytes,
    required this.status,
    this.hlsKey,
  });

  final String id;
  final Uri uri;
  final int priority;
  final double progress;
  final int cachedBytes;
  final int downloadedBytes;
  final int totalBytes;
  final frb_status.DownloadStatus status;
  final String? hlsKey;

  factory DownloadTask.fromInfo(frb.DownloadTaskInfo info) => DownloadTask(
    id: info.id,
    uri: Uri.parse(info.url),
    priority: info.priority,
    progress: info.progress,
    cachedBytes: info.cachedBytes.toInt(),
    downloadedBytes: info.downloadedBytes.toInt(),
    totalBytes: info.totalBytes.toInt(),
    status: info.status,
    hlsKey: info.hlsKey,
  );

  String get url => uri.toString();
}

/// Download manager facade backed by the Rust download pool.
class DownloadManager {
  Stream<DownloadTask> get stream =>
      frb.downloadManagerSubscribe().map(DownloadTask.fromInfo);

  Future<List<DownloadTask>> get allTasks async =>
      (await frb.downloadManagerAllTasks()).map(DownloadTask.fromInfo).toList();

  Future<List<DownloadTask>> get downloadingTasks async =>
      (await frb.downloadManagerDownloadingTasks())
          .map(DownloadTask.fromInfo)
          .toList();

  Future<void> pauseTaskById(String taskId) =>
      frb.downloadManagerPauseTaskById(taskId: taskId);

  Future<void> resumeTaskById(String taskId) =>
      frb.downloadManagerResumeTaskById(taskId: taskId);

  Future<void> cancelTaskById(String taskId) =>
      frb.downloadManagerCancelTaskById(taskId: taskId);

  Future<void> cancelTaskByUrl(String url) =>
      frb.downloadManagerCancelTaskByUrl(url: url);

  Future<void> pauseTaskByUrl(String url) =>
      frb.downloadManagerPauseTaskByUrl(url: url);

  Future<void> resumeTaskByUrl(String url) =>
      frb.downloadManagerResumeTaskByUrl(url: url);

  Future<void> pauseAllTasks() => frb.downloadManagerPauseAllTasks();

  Future<void> cancelAllTask() => frb.downloadManagerCancelAllTasks();

  Future<void> cancelTaskAboutUrl(String url) =>
      frb.downloadManagerCancelTaskAboutUrl(url: url);

  void dispose() {}
}

File: rust/src/api/download_manager.rs
use std::sync::Arc;

use crate::frb_generated::StreamSink;
use parking_lot::Mutex;

use crate::download::{DownloadStatus, DownloadTask};
use crate::proxy::require_runtime;

#[flutter_rust_bridge::frb]
#[derive(Debug, Clone)]
pub struct DownloadTaskInfo {
    pub id: String,
    pub url: String,
    pub priority: i32,
    pub progress: f64,
    pub cached_bytes: i64,
    pub downloaded_bytes: i64,
    pub total_bytes: i64,
    pub status: DownloadStatus,
    pub hls_key: Option<String>,
}

fn task_to_info(task: &DownloadTask) -> DownloadTaskInfo {
    DownloadTaskInfo {
        id: task.id.clone(),
        url: task.url(),
        priority: task.priority,
        progress: task.progress,
        cached_bytes: task.cached_bytes,
        downloaded_bytes: task.downloaded_bytes,
        total_bytes: task.total_bytes,
        status: task.status,
        hls_key: task.hls_key.clone(),
    }
}

fn tasks_to_info(tasks: Vec<Arc<Mutex<DownloadTask>>>) -> Vec<DownloadTaskInfo> {
    tasks.iter().map(|t| task_to_info(&t.lock())).collect()
}

#[flutter_rust_bridge::frb]
pub async fn download_manager_subscribe(sink: StreamSink<DownloadTaskInfo>) -> Result<(), String> {
    let runtime = require_runtime()?;
    let mut rx = runtime.downloads().subscribe();
    tokio::spawn(async move {
        loop {
            match rx.recv().await {
                Ok(task) => {
                    let info = task_to_info(&task.lock());
                    if sink.add(info).is_err() {
                        break;
                    }
                }
                Err(tokio::sync::broadcast::error::RecvError::Closed) => break,
                Err(tokio::sync::broadcast::error::RecvError::Lagged(_)) => continue,
            }
        }
    });
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_all_tasks() -> Result<Vec<DownloadTaskInfo>, String> {
    let runtime = require_runtime()?;
    Ok(tasks_to_info(runtime.downloads().task_list()))
}

#[flutter_rust_bridge::frb]
pub fn download_manager_downloading_tasks() -> Result<Vec<DownloadTaskInfo>, String> {
    let runtime = require_runtime()?;
    Ok(tasks_to_info(runtime.downloads().downloading_tasks()))
}

#[flutter_rust_bridge::frb]
pub fn download_manager_pause_task_by_id(task_id: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().pause_task_by_id(&task_id);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_resume_task_by_id(task_id: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().resume_task_by_id(&task_id);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_cancel_task_by_id(task_id: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().cancel_task_by_id(&task_id);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_pause_all_tasks() -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().pause_all_tasks();
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_cancel_all_tasks() -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().cancel_all_tasks();
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_cancel_task_by_url(url: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().cancel_task_by_url(&url);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_pause_task_by_url(url: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().pause_task_by_url(&url);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub fn download_manager_resume_task_by_url(url: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().resume_task_by_url(&url);
    Ok(())
}

#[flutter_rust_bridge::frb]
pub async fn download_manager_cancel_task_about_url(url: String) -> Result<(), String> {
    let runtime = require_runtime()?;
    runtime.downloads().cancel_task_about_url(&url);
    Ok(())
}
